src/paillier.rs

```diff
@@ -131,50 +131,62 @@ pub(crate) mod prime_gen {
     use rand::Rng;
     use rand::{CryptoRng, RngCore};
 
-    /// Sample a 512-bit safe prime uniformly at random.
-    ///
-    /// Prime size is derived from the security
-    /// parameter setting of κ = 128, and safe primes being of length 4κ (Figure 6,
-    /// Round 1 of the CGGMP'21 paper)
-    pub(crate) fn get_random_safe_prime_512<R: RngCore + CryptoRng>(rng: &mut R) -> BigNumber {
+    /// Sample a safe prime with length `PRIME_BITS` at random.
+    pub(crate) fn get_random_safe_prime<R: RngCore + CryptoRng>(rng: &mut R) -> BigNumber {
         BigNumber::safe_prime_from_rng(PRIME_BITS, rng)
     }
 
-    // Generate safe primes from a file. Usually, generating safe primes takes
-    // awhile (0-5 minutes per 512-bit safe prime on my laptop, average 50 seconds)
     #[cfg(test)]
     lazy_static::lazy_static! {
+        /// List of `PRIME_BITS`-length safe primes, generated _insecurely_.
         static ref POOL_OF_PRIMES: Vec<BigNumber> = get_safe_primes_from_file();
     }
 
+    /// Load a set of pre-generated safe primes from a file for testing efficiency.
     #[cfg(test)]
     fn get_safe_primes_from_file() -> Vec<BigNumber> {
-        // The currently-generated file includes safe primes of different lengths (511-514), so
-        // we filter them to get only the ones exactly `PRIME_BITS` long.
-        crate::safe_primes_512::SAFE_PRIMES
-            .iter()
-            .map(|s| BigNumber::from_slice(hex::decode(s).unwrap()))
-            .filter(|prime| prime.bit_length() == PRIME_BITS)
-            .collect()
+        match PRIME_BITS {
+            // The list of 512-bit primes includes safe primes of different lengths (511-514), so
+            // we filter out any that aren't exactly 512.
+            512 => crate::safe_primes_512::SAFE_PRIMES
+                .iter()
+                .map(|s| BigNumber::from_slice(hex::decode(s).unwrap()))
+                .filter(|prime| prime.bit_length() == PRIME_BITS)
+                .collect(),
+
+            // This is the recommended `PRIME_BITS` value.
+            1024 => crate::safe_primes_1024::SAFE_PRIMES
+                .iter()
+                .map(|s| BigNumber::from_slice(hex::decode(s).unwrap()))
+                .collect(),
+
+            _ => Vec::new(),
```

Comment:
Is it worth replacing this with a more explicit error? Otherwise, changing PRIME_BITS causes tests to fail in a more opaque way (thread 'my_test_name' panicked at 'cannot sample empty range')

Reply:
I made some changes and now tests that call `get_prime_pair_from_pool_insecure` directly get a more meaningful error, but those that call `PaillierDecryptionKey::new` fail with a `RetryFailed` error. I'm going to say that gives enough info for now, but we might revisit in the future.

Good catch, though. I didn't realize the `rng` function panics with a bad range.